src/nnet3/discriminative-training.h

```diff
@@ -0,0 +1,244 @@
+  // the objf will be -0.5 times this constant times the squared l2 norm.
+  // (squared so it's additive across the dimensions).  e.g. try 0.0005.
+  BaseFloat l2_regularize;
+
+  DiscriminativeTrainingOptions(): criterion("smbr"), 
+                                   acoustic_scale(0.1),
+                                   drop_frames(false),
+                                   one_silence_class(false),
+                                   boost(0.0), 
+                                   xent_regularize(0.0), 
+                                   l2_regularize(0.0) { }
+
+  void Register(OptionsItf *opts) {
+    opts->Register("criterion", &criterion, "Criterion, 'mmi'|'mpfe'|'smbr', "
+                   "determines the objective function to use.  Should match "
+                   "option used when we created the examples.");
+    opts->Register("acoustic-scale", &acoustic_scale, "Weighting factor to "
+                   "apply to acoustic likelihoods.");
+    opts->Register("drop-frames", &drop_frames, "For MMI, if true we drop frames "
+                   "with no overlap of num and den pdf-ids");
+    opts->Register("boost", &boost, "Boosting factor for boosted MMI (e.g. 0.1)");
+    opts->Register("one-silence-class", &one_silence_class, "If true, newer "
+                   "behavior which will tend to reduce insertions "
+                   "when using MPFE or SMBR objective");
+    opts->Register("silence-phones", &silence_phones_str,
+                   "For MPFE or SMBR objectives, colon-separated list of "
+                   "integer ids of silence phones, e.g. 1:2:3");
+    opts->Register("l2-regularize", &l2_regularize, "l2 regularization "
+                   "constant for 'chain' output "
+                   "of the neural net.");
+    opts->Register("xent-regularize", &xent_regularize, "Cross-entropy "
+                   "regularization constant for sequence training.  If "
+                   "nonzero, the network is expected to have an output "
+                   "named 'output-xent', which should have a softmax as "
+                   "its final nonlinearity.");
+  }
+};
+
+struct DiscriminativeTrainingStatsOptions {
+  bool accumulate_gradients;
+  bool accumulate_output;
+  int32 num_pdfs;
+
+  void Register(OptionsItf *opts) {
+    opts->Register("accumulate-gradients", &accumulate_gradients,
+                   "Accumulate gradients for debugging discriminative training");
+    opts->Register("accumulate-output", &accumulate_output,
+                   "Accumulate nnet output "
+                   "for debugging discriminative training");
+    opts->Register("num-pdfs", &num_pdfs,
+                   "Number of pdfs");
+  }
+
+  DiscriminativeTrainingStatsOptions() :
+    accumulate_gradients(false), accumulate_output(false),
+    num_pdfs(0) { }
+};
+
+struct DiscriminativeTrainingStats {
+  double tot_t;          // total number of frames
```

Comment:
DiscriminativeTrainingStats is not a good name because it conjures up sufficient statistics, which would imply this is the main thing you're accumulating.  You could call it DiscriminativeObjectiveInfo.